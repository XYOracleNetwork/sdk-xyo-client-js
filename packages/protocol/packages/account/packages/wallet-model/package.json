{
  "name": "@xyo-network/wallet-model",
  "author": {
    "email": "support@xyo.network",
    "name": "XYO Development Team",
    "url": "https://xyo.network"
  },
  "bugs": {
    "email": "support@xyo.network",
    "url": "https://github.com/XYOracleNetwork/sdk-xyo-client-js/issues"
  },
  "description": "Primary SDK for using XYO Protocol 2.0",
  "dependencies": {
    "@xylabs/hex": "^4.0.0",
    "@xyo-network/account-model": "workspace:^"
  },
  "devDependencies": {
    "@xylabs/ts-scripts-yarn3": "^4.0.0-rc.15",
    "@xylabs/tsconfig": "^4.0.0-rc.15",
    "ethers": "^6.13.2",
    "typescript": "^5.5.4"
  },
  "peerDependencies": {
    "ethers": "^6"
  },
  "peerDependenciesMeta": {
    "ethers": {
      "optional": true
    }
  },
  "exports": {
    ".": {
      "types": "./dist/neutral/index.d.ts",
      "default": "./dist/neutral/index.mjs"
    },
    "./package.json": "./package.json"
  },
  "module": "dist/neutral/index.mjs",
  "types": "dist/neutral/index.d.ts",
  "homepage": "https://xyo.network",
  "license": "LGPL-3.0-only",
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/XYOracleNetwork/sdk-xyo-client-js.git"
  },
  "sideEffects": false,
  "version": "3.0.0",
  "type": "module"
}
