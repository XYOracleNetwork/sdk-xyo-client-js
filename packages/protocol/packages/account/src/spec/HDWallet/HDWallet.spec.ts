import { WalletStatic } from '@xyo-network/wallet-model'

export const generateHDWalletTests = (title: string, HDWallet: WalletStatic) => {
  describe(title, () => {
    const mnemonic = 'later puppy sound rebuild rebuild noise ozone amazing hope broccoli crystal grief'
    describe('constructor', () => {
      it('can be created from mnemonic', async () => {
        const sut = await HDWallet.fromPhrase(mnemonic)
        expect(sut).toBeDefined()
      })
    })
    describe('derivePath', () => {
      const paths = ['0/4', "44'/0'/0'", "44'/60'/0'/0/0", "44'/60'/0'/0/1", "49'/0'/0'", "84'/0'/0'", "84'/0'/0'/0"]
      it.each(paths)('works repeatably & interoperably', async (path: string) => {
        const sutA = await HDWallet.fromPhrase(mnemonic)
        const sutB = await HDWallet.fromExtendedKey(sutA.extendedKey)
        const accountA = await sutA.derivePath?.(path)
        const accountB = await sutB.derivePath?.(path)
        // expect(accountA.path).toBe(accountB.path)
        expect(accountA.address).toBe(accountB.address)
        expect(accountA.private.hex).toBe(accountB.private.hex)
        expect(accountA.public.hex).toBe(accountB.public.hex)
        expect(accountA.address).toMatchSnapshot()
        expect(accountB.address).toMatchSnapshot()
      })
      it('works when paths provided incrementally', async () => {
        const parent = "44'/60'/0'"
        const child = '0/1'
        const sutA = await HDWallet.fromPhrase(mnemonic)
        const sutB = await HDWallet.fromPhrase(mnemonic)
        expect(sutA.path).toEqual(sutB.path)
        const accountA = await (await sutA.derivePath(parent)).derivePath?.(child)
        const accountB = await sutB.derivePath?.([parent, child].join('/'))
        expect(accountA.address).toEqual(accountB.address)
        expect(accountA.private.hex).toEqual(accountB.private.hex)
        expect(accountA.public.hex).toEqual(accountB.public.hex)
        expect(accountA.address).toMatchSnapshot()
        expect(accountB.address).toMatchSnapshot()
      })
      it('works when paths provided absolutely', async () => {
        const parent = "44'/60'/0'"
        const child = '0/1'
        const sutA = await HDWallet.fromPhrase(mnemonic)
        const sutB = await HDWallet.fromPhrase(mnemonic)
        expect(sutA.path).toEqual(sutB.path)
        const accountA = await (await sutA.derivePath(parent)).derivePath?.(child)
        const accountB = await sutB.derivePath?.(['m', parent, child].join('/'))
        expect(accountA.address).toEqual(accountB.address)
        expect(accountA.private.hex).toEqual(accountB.private.hex)
        expect(accountA.public.hex).toEqual(accountB.public.hex)
        expect(accountA.address).toMatchSnapshot()
        expect(accountB.address).toMatchSnapshot()
      })
      it('returns cached instances on subsequent requests', async () => {
        const parent = "44'/60'/0'"
        const child = '0/1'
        const sutA = await HDWallet.fromPhrase(mnemonic)
        const sutB = await HDWallet.fromPhrase(mnemonic)
        // sutA and sutB should be the same instance
        expect(sutA).toBe(sutB)
        const accountA = await (await sutA.derivePath(parent)).derivePath?.(child)
        const accountB = await sutB.derivePath?.([parent, child].join('/'))
        expect(accountA.address).toEqual(accountB.address)
        expect(accountA.private.hex).toEqual(accountB.private.hex)
        expect(accountA.public.hex).toEqual(accountB.public.hex)
        // accountA and accountB should be the same instance
        // expect(accountA).toBe(accountB)
      })
    })
  })
}

test('HDWallet tests generator is defined', () => {
  expect(generateHDWalletTests).toBeFunction()
})
