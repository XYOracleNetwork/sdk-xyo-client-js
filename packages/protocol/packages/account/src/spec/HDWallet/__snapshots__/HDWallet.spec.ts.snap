// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Same address, two paths 1`] = `"1eefd4fb703a43ce5ded91bdff2820a7a3999ca8"`;

exports[`Same address, two paths 2`] = `"1b9eedc3466de680d07b1e14ae69281c14fa47be"`;

exports[`Same address, two paths 3`] = `"1b9eedc3466de680d07b1e14ae69281c14fa47be"`;

exports[`Same address, two paths 4`] = `"e34188f2e84688b036f47bc2ad57aca1085cc95a"`;
